Explorer/Assets/DCL/Landscape/TerrainGenerator.cs

```diff
@@ -350,203 +342,156 @@ private static Texture2D CreateOccupancyMap(NativeParallelHashSet<int2> ownedPar
         }
 
-        private static int WriteInteriorChamferOnWhite(Texture2D r8, int2 minParcel, int2 maxParcel, int padding)
+        internal static (int floor, int maxSteps) WriteInteriorChamferOnWhite(Texture2D texture, int2 minParcel, int2 maxParcel, int padding)
         {
-            int w = r8.width, h = r8.height, n = w * h;
-            NativeArray<byte> src = r8.GetRawTextureData<byte>();
-            if (!src.IsCreated || src.Length != n) return 0;
-
-            int textureHalfSize = w / 2; // Assuming square texture
-
-            // Calculate working area bounds in texture coordinates (with padding)
-            int workMinX = minParcel.x - padding;
-            int workMinY = minParcel.y - padding;
-            int workMaxX = maxParcel.x + padding;
-            int workMaxY = maxParcel.y + padding;
-
-            // Clamp to texture logical range [-textureHalfSize .. textureHalfSize-1]
-            workMinX = Math.Max(workMinX, -textureHalfSize);
-            workMinY = Math.Max(workMinY, -textureHalfSize);
-            workMaxX = Math.Min(workMaxX, textureHalfSize - 1);
-            workMaxY = Math.Min(workMaxY, textureHalfSize - 1);
-
-            // Convert to texture pixel coordinates and expand by 1 to include border stripes
-            int texMinX = workMinX + textureHalfSize - 1;
-            int texMinY = workMinY + textureHalfSize - 1;
-            int texMaxX = workMaxX + textureHalfSize + 1;
-            int texMaxY = workMaxY + textureHalfSize + 1;
-
-            // Clamp to actual texture bounds
-            texMinX = Math.Max(texMinX, 0);
-            texMinY = Math.Max(texMinY, 0);
-            texMaxX = Math.Min(texMaxX, w - 1);
-            texMaxY = Math.Min(texMaxY, h - 1);
-
-            const int INF = 1 << 28;
-            const int ORTH = 3; // 3-4 chamfer (good Euclidean approx)
-            const int DIAG = 4;
+            NativeArray<byte> src = texture.GetRawTextureData<byte>();
+            if (!src.IsCreated) return (0, 0);
+
+            int textureWidth = texture.width;
+
+            if (!ispow2(textureWidth))
+                ReportHub.LogError(ReportCategory.LANDSCAPE, "OccupancyMap is not power of 2. This can lead to unpredictable results!");
+
+            // Calculate working area bounds
+            int textureHalfSize = textureWidth / 2;
+            int minX = max(minParcel.x - padding + textureHalfSize, 0);
+            int minY = max(minParcel.y - padding + textureHalfSize, 0);
+            int maxX = min(maxParcel.x + padding + textureHalfSize, textureWidth - 1);
+            int maxY = min(maxParcel.y + padding + textureHalfSize, textureWidth - 1);
+
+            var workingArea = new RectInt(minX, minY, maxX - minX, maxY - minY);
 
             // Seed distances at BLACK pixels (occupied parcels - leave them 0), propagate into WHITE (free parcels)
-            var dist = new int[n];
-            bool anyBlack = false, anyWhite = false;
+            var dist = new int[src.Length];
```

Comment:
You probably could avoid allocation by using a temp `NativeArray` or allocation on a stack if the size of the array is controllable and small